src/components/Map.js

```diff
@@ -1,229 +1,196 @@
-		this.leafletMarkerLayers = {};
-		this.leafletPolylineLayers = {};
-		this.leafletHeatmapLayers = {};
-		this.leafletClusterLayers = {};
-		this.layerControl = null;
-	}
-
-
-	componentDidUpdate() {
-		let layers = this.props.layers;
-		let globalBounds = new L.LatLngBounds();
-		let ukeyMarkerArray = [];
-		let ukeyPolylineArray = [];
-		let ukeyHeatmapArray = [];
-		let ukeyClusterArray = [];
-		// Iterate through layers
-		layers.map((layer) => {
-			if (layer.ukey === undefined)
-				return null;
-			let bds = new L.LatLngBounds(layer.bounds);
-			if (bds.isValid())
-				globalBounds.extend(bds);
-			if (layer.rendering === RENDERING_MARKERS) {
-				ukeyMarkerArray.push(layer.ukey);
-				if (!this.leafletMarkerLayers[layer.ukey]) {
-					this.leafletMarkerLayers[layer.ukey] = L.layerGroup().addTo(this.map);
-				}
-				this.updateMarkerLayer(layer.data, layer.color, layer.ukey);
-			} else if (layer.rendering === RENDERING_POLYLINE) {
-				ukeyPolylineArray.push(layer.ukey);
-				if (this.leafletPolylineLayers[layer.ukey]) {
-					// todo find a way of updating the polyline layer instead of delete & recreate
-					this.map.removeLayer(this.leafletPolylineLayers[layer.ukey]);
+	React.useEffect(() => {
+		const map = mapRef.current;
+
+		if (map) {
+
+			layerControlRef.current.remove(map);
+
+			let globalBounds = new L.LatLngBounds();
+
+			
+			// On a new render pass, build new map overlays object,
+			// and replace the current map overlays object. This auto-removes
+			// any deleted layers from map
+			const newMapOverlays = {}
+			const currentmapOverlays  = mapOverlaysRef.current;
+
+			// Iterate through layers
+			layers.forEach((layer) => {
+				const {ukey, name, bounds, rendering, data, radius, color } = layer;
+
+				if (ukey === undefined) {
+					return;
 				}
-				this.updatePolylineLayer(layer.data, layer.color, layer.ukey);
-			} else if (layer.rendering === RENDERING_HEATMAP) {
-				ukeyHeatmapArray.push(layer.ukey);
-				if (this.leafletHeatmapLayers[layer.ukey]) {
-					// todo find a way of updating the heat layer instead of delete & recreate
-					this.map.removeLayer(this.leafletHeatmapLayers[layer.ukey]);
+
+				const latLngBounds = new L.LatLngBounds(bounds);
+
+				if (latLngBounds.isValid()) {
+					globalBounds.extend(latLngBounds);
 				}
-				this.updateHeatmapLayer(layer.data, layer.radius, layer.ukey);
-			} else if (layer.rendering === RENDERING_CLUSTERS) {
-				ukeyClusterArray.push(layer.ukey);
-				if (!this.leafletClusterLayers[layer.ukey]) {
-					this.leafletClusterLayers[layer.ukey] = L.markerClusterGroup();
-					this.map.addLayer(this.leafletClusterLayers[layer.ukey]);
+
+				const rgbColor = `rgb(${color.r}, ${color.g}, ${color.b})`;
+
+				switch(rendering) {
+					case RENDERING_MARKERS:
+						let markerLayer = currentmapOverlays[name];
+
+						if (!markerLayer) {
+							markerLayer = L.layerGroup().addTo(map);
+						}
+
+						// TODO: check if the layer has changed before rerendering it
```

Comment:
Would that be possible (or relevant?) with this useEffect hook?

Reply:
It's possible to write a data comparator for the layer. Given the volume of data / layer though , I don't yet see a path to an efficient comparator and it may be okay to just refresh data in the resulting overlay

Reply:
Ok, as a first step we could save in the layer whether its configuration changed or not. If it changed, then it is impossible to determine if the new configuration lead to the same data, but if it is unchanged then we can just : 

1. re-render if the rendering type has changed
2. use the old one if it is unchanged

Just some ideas for later ;-)